pytensor/tensor/extra_ops.py

```diff
@@ -1224,60 +1223,37 @@ def make_node(self, x):
         return Apply(self, [x], outputs)
 
     def perform(self, node, inputs, output_storage):
-        x = inputs[0]
-        z = output_storage
-        param = {}
-        if self.return_index:
-            param["return_index"] = True
-        if self.return_inverse:
-            param["return_inverse"] = True
-        if self.return_counts:
-            param["return_counts"] = True
-        if self.axis is not None:
-            param["axis"] = self.axis
-        outs = np.unique(x, **param)
-        if (
-            (not self.return_inverse)
-            and (not self.return_index)
-            and (not self.return_counts)
-        ):
-            z[0][0] = outs
-        else:
+        [x] = inputs
+        outs = np.unique(
+            x,
+            return_index=self.return_index,
+            return_inverse=self.return_inverse,
+            return_counts=self.return_counts,
+            axis=self.axis,
+        )
+        if isinstance(outs, tuple):
             for i in range(len(outs)):
-                z[i][0] = outs[i]
+                output_storage[i][0] = outs[i]
+        else:
+            output_storage[0][0] = outs
 
     def infer_shape(self, fgraph, node, i0_shapes):
-        ret = fgraph.shape_feature.default_infer_shape(fgraph, node, i0_shapes)
-        if self.axis is not None:
-            self_axis = self.axis
-            ndim = len(i0_shapes[0])
-            if self_axis < 0:
-                self_axis += ndim
-            if self_axis < 0 or self_axis >= ndim:
-                raise RuntimeError(
-                    f"Unique axis `{self.axis}` is outside of input ndim = {ndim}."
-                )
-            ret[0] = tuple(
-                fgraph.shape_feature.shape_ir(i, node.outputs[0]) for i in range(ndim)
-            )
+        [x_shape] = i0_shapes
```

Comment:
I am not sure I understand what is happening in this function, but just to check, shouldn't there be a case for `return_index` and `return_counts` as well?

Reply:
There is. `i0_shapes` are the input dimensions, so that doesn't change with the number of outputs. `return_index`/`counts` are outputs, and they are always vector.

We set `out_shapes = [out.shape[0] for out in node.outputs]` by default which will always work for `return_index` and `return_counts`. Then we have special logic for the main output when axis is not None and for `return_inverse` which is not just `out.shape[0]`.

Reply:
The big picture is the function tries to return a graph for the shape of the outputs given the input shapes and (possible values, which you could retrieve from node.inputs). The default graph of the shape is just output.shape, which we try to avoid when possible, as we would like to avoid computing the Op in order to find out its shape. For unique we can do that for some of the outputs dimensions, but not all (we only know how many repeated values there are if we evaluate `Unique`. This function is combining dims we can know from the input shapes and those that we can only get after we compute the outputs `with out.shape[0]`.